### PR TITLE
 Adding baseball hits examples using NUTS and Inference utils

### PR DESCRIPTION
**Pull request by neerajprad**
This also contains the changes in #1079.

### CODE REVIEW

File: examples/baseball.py [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+    Number of hits has a Binomial distribution with a logit link function.
+    The logits $\alpha$ for each player is normally distributed with the
+    mean and scale parameters sharing a common prior.
+
+    :param (torch.Tensor) at_bats: Number of at bats for each player.
+    :return: Number of hits predicted by the model.
+    """
+    num_players = at_bats.shape[0]
+    loc = pyro.sample("loc", Normal(at_bats.new_tensor(-1), at_bats.new_tensor(1)))
+    scale = pyro.sample("scale", HalfCauchy(at_bats.new_tensor(0), at_bats.new_tensor(1)))
```
Comment (martinjankowiak) [inline examples/baseball.py]: why HalfCauchy instead of TruncNormal?

Comment (fehiepsi) [reply examples/baseball.py]: @martinjankowiak  Although the Stan tutorial used HalfNormal,  I think that HalfNormal is not available yet in Pytorch and Pyro (not so sure). And I think that people use HalfCauchy more often.

**Comment by fehiepsi** [comment]
Thanks @neerajprad ! I will make a review soon.

File: examples/baseball.py [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+    """
+    Number of hits in $K$ at bats for each player has a Binomial
+    distribution with a common probability of success, $\phi$.
+
+    :param (torch.Tensor) at_bats: Number of at bats for each player.
+    :return: Number of hits predicted by the model.
+    """
+    phi_prior = Uniform(at_bats.new_tensor(0), at_bats.new_tensor(1))
+    phi = pyro.sample("phi", phi_prior)
+    return pyro.sample("obs", Binomial(at_bats, phi))
```
[the remark this answers is not in the document]
Comment (eb8680) [reply examples/baseball.py]: Do you mean subsampling?

[the remark this answers is not in the document]
Comment (fritzo) [reply examples/baseball.py]: IIUC `iarange` would not help. First HMC is incompatible with subsampling. Second, without subsampling the only benefit provided by `iarange` is to declare independence. This declaration helps eliminate zero-expectation ELBO terms in SVI, but it is not used in HMC where we simply compute the total `log_prob` and its gradient.

[the remark this answers is not in the document]
Comment (fehiepsi) [reply examples/baseball.py]: Good to know that! Now I somehow understand what you mentioned a while ago: use `iarange` instead of independent will reduce variance thanks to those eliminating jobs of iarange. Thanks @fritzo !

**Comment by neerajprad** [comment]
> All the models look good to me. If possible, we can also implement "Model 3: Partial Pooling (Chance of Success)". @neerajprad I guess you wonder that there is no Pareto distribution available to Pyro/Pytorch. 

@fehiepsi - I think `Pareto` is available in PyTorch. :) Do you mind if we implement this later? It doesn't perform as well as the logit model, and I was having some problems with stability when I tried it earlier.

File: examples/baseball.py [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+def not_pooled(at_bats):
+    """
+    Number of hits in $K$ at bats for each player has a Binomial
+    distribution with independent probability of success, $\phi_i$.
+
+    :param (torch.Tensor) at_bats: Number of at bats for each player.
+    :return: Number of hits predicted by the model.
+    """
+    num_players = at_bats.shape[0]
+    phi_prior = Uniform(at_bats.new_tensor(0), at_bats.new_tensor(1)).expand_by([num_players]).independent(1)
```
[the remark this answers is not in the document]
Comment (neerajprad) [reply examples/baseball.py]: @fehiepsi - I can confirm that `Uniform(at_bats.new_zeros(num_players), at_bats.new_ones(num_players))` is identical. What were you getting for GP, I am curious too? :)

File: examples/baseball.py [hunk 4]
```diff
@@ -0,0 +1,275 @@
+from __future__ import absolute_import, division, print_function
+
+import argparse
+import logging
+import math
+
+import numpy as np
+import pandas as pd
```
Comment (jpchen) [inline examples/baseball.py]: can you add `pandas` to the `extra_requires` in `setup.py`

Comment (neerajprad) [reply examples/baseball.py]: I have added it; and its being tested in test-examples.
